Kuroba/app/src/main/java/com/github/adamantcheese/chan/report/ReportProblemController.kt

```diff
@@ -0,0 +1,154 @@
+
+    fun setOnFinishedCallback(callback: () -> Unit) {
+        this.onFinishedListener = callback
+    }
+
+    private fun onSendReportClick() {
+        val title = reportActivityProblemTitle.text?.toString() ?: ""
+        val description = reportActivityProblemDescription.text?.toString() ?: ""
+        val logs = reportActivityLogsText.text?.toString() ?: ""
+
+        if (title.isEmpty()) {
+            reportActivityProblemTitle.error =
+                    context.getString(R.string.report_activity_title_cannot_be_empty_error)
+            return
+        }
+
+        if (description.isEmpty()) {
+            reportActivityProblemDescription.error =
+                    context.getString(R.string.report_activity_description_cannot_be_empty_error)
+            return
+        }
+
+        if (reportActivityAttachLogsButton.isChecked && logs.isEmpty()) {
+            reportActivityLogsText.error =
+                    context.getString(R.string.report_activity_logs_are_empty_error)
+            return
+        }
+
+        val logsParam = if (!reportActivityAttachLogsButton.isChecked) {
+            null
+        } else {
+            logs
+        }
+
+        showProgressDialog()
+
+        presenter.sendReport(title, description, logsParam)
+                .subscribeOn(Schedulers.io())
+                .observeOn(AndroidSchedulers.mainThread())
+                .doOnTerminate { hideProgressDialog() }
+                .subscribe({ result ->
+                    handleResult(result)
+                }, { error ->
+                    Logger.e(TAG, "Error", error)
+
+                    val errorMessage = error.message ?: "No error message"
+                    AndroidUtils.showToast(
+                            context.getString(R.string.report_activity_error_while_trying_to_send_report,
+                                    errorMessage)
+                    )
+                })
+                .also { disposable -> compositeDisposable.add(disposable) }
+    }
+
+    private fun handleResult(result: MResult<Boolean>) {
+        when (result) {
+            is MResult.Value -> {
+                AndroidUtils.showToast(context.getString(R.string.report_activity_report_sent_message))
+                onFinishedListener?.invoke()
+            }
+            is MResult.Error -> {
+                val errorMessage = result.error.message ?: "No error message"
+                AndroidUtils.showToast(
+                        context.getString(R.string.report_activity_error_while_trying_to_send_report,
+                                errorMessage)
+                )
+            }
+        }
+    }
+
+    private fun showProgressDialog() {
+        hideProgressDialog()
+
+        loadingViewController = LoadingViewController(context, true)
+        presentController(loadingViewController)
+    }
+
+    private fun hideProgressDialog() {
+        if (loadingViewController != null) {
+            loadingViewController!!.stopPresenting()
```

Comment:
Do you need to do the null check here again? You're checking it in the if statement already, and it would be really poor ironic to throw an exception when trying to report a bug.

Reply:
loadingViewController is a shared mutable nullable variable. Kotlin cannot guarantee that it won't be changed in a different place (different thread) so you have to always treat it like nullable. But there is a more elegant way to do this, and I should probably use it.

Reply:
Done.